torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp

```diff
@@ -511,6 +552,98 @@ void InsertQuantDeQuantHelper::cleanup(Module& module, Graph* g) {
   GRAPH_DUMP("After remove observers :", g);
 }
 
+Value* ProcessWeightObserverSubgraph::updateInputValueInGraph(
+    Value* v,
+    std::shared_ptr<Graph>& g) {
+  if (remap_values_in_observer_subgraph_.count(v) == 0) {
+    auto new_value = g->block()->addInput();
+    remap_values_in_observer_subgraph_[v] = new_value;
+    new_value->copyMetadata(v);
+    return new_value;
+  } else {
+    return remap_values_in_observer_subgraph_[v];
+  }
+}
+
+void ProcessWeightObserverSubgraph::cloneNodeInGraph(
+    Node* node,
+    std::shared_ptr<Graph>& g) {
+  auto* block = g->block();
+  auto value_fn = [&](Value* v) { return updateInputValueInGraph(v, g); };
+
+  auto new_node = block->appendNode(g->createClone(node, value_fn));
+  for (size_t i = 0; i < node->outputs().size(); ++i) {
+    auto oo = node->outputs()[i];
+    auto no = new_node->outputs()[i];
+    remap_values_in_observer_subgraph_[oo] = no;
+  }
+}
+void ProcessWeightObserverSubgraph::buildObserverSubgraph(
+    const std::vector<Node*>& weight_subgraph,
+    std::shared_ptr<Graph> dest_graph) {
+  // Build weight subgraph
+  for (auto n : weight_subgraph) {
+    cloneNodeInGraph(n, dest_graph);
+  }
+  LintGraph(dest_graph);
+  // Add last node output value as subgraph output.
+  dest_graph->registerOutput(
+      updateInputValueInGraph(weight_subgraph.back()->output(0), dest_graph));
```

Comment:
we can make this more general: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/ir/ir.cpp.html#661

Reply:
actually, is it possible to clone the graph with: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/ir/ir.cpp.html#_ZN5torch3jit5Graph11createCloneEPNS0_4NodeERKSt8functionIFPNS0_5ValueES6_EEb

Reply:
@jerryzh168 I do use `createClone` please see https://github.com/pytorch/pytorch/pull/39412/files#diff-901c179fa606354f6bc42223debab96cR574

Reply:
I see. I looked at the code base and looks like the patterns are a little bit different: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/canonicalize.cpp.html#_ZN5torch3jit12CanonicalizeERKSt10shared_ptrINS0_5GraphEEb, could you follow the same pattern as the existing code?

Reply:
Also, should we check that the graph only has one input(%self)?

Reply:
As long as the other input doesn't contribute to the weight observer we should be fine. I will add an assert for that in `findSubgraph`

Reply:
> As long as the other input doesn't contribute to the weight observer we should be fine. I will add an assert for that in `findSubgraph`

to clarify we should only accept the graph with one input since when we run the model we pass in a single input, maybe we should also check the input is %self as well, that is, make sure the type of the input matches the module.

Reply:
This case should be fine imo as long as the extra inputs don't produce the weight, they wont be in the new graph.
Original graph 
```
graph(%self : __torch__.___torch_mangle_129.DynamicModel,
      %x.1 : Tensor,
      %x.2: Tensor):
...
%6 : Tensor = prim::GetAttr[name="weight"](%self)
..(%x.1 and %x.2 are used here but don't affect weight.
```
New graph
```
graph(%self : __torch__.___torch_mangle_129.DynamicModel):
...
%6 : Tensor = prim::GetAttr[name="weight"](%self)
(code to run observer forward)
```

Reply:
We have different graphs in mind, I'm talking about the extracted subgraph.